htp/htp_transaction.c

```diff
@@ -789,6 +789,25 @@ static htp_status_t htp_tx_res_process_body_data_decompressor_callback(htp_tx_da
     // Invoke all callbacks.
     htp_status_t rc = htp_res_run_hook_body_data(d->tx->connp, d);
     if (rc != HTP_OK) return HTP_ERROR;
+    d->tx->connp->out_decompressor->nb_callbacks++;
+    if (d->tx->connp->out_decompressor->nb_callbacks % HTP_COMPRESSION_TIME_FREQ_TEST == 0) {
+        struct timeval after, duration;
+        gettimeofday(&after, NULL);
+        // sanity check for race condition if system time changed
+        if ( timercmp(&after, &d->tx->connp->out_decompressor->time_before, >) ) {
+            timersub(&after, &d->tx->connp->out_decompressor->time_before, &duration);
+            timeradd(&d->tx->connp->out_decompressor->time_spent, &duration, &d->tx->connp->out_decompressor->time_spent);
```

Comment:
is there a reason for using `duration` here, instead of just calculating `time_spent` directly from `time_before` and `after`?

Reply:
Yes, cumulate time spent over multiple calls to `decompress` for one response
This may be over cautious... But I was worried to forget one corner case.
What do you think ?

Reply:
Maybe I am misunderstanding the overall flow. But I see that `htp_tx_res_process_body_data_ex` we set the `time_before` and then after running the decompress check if it took too long. We do this by storing the diff between before and after in the local variable `duration`, which is then added to `time_spent`.

But then in `htp_tx_res_process_body_data_decompressor_callback`, which I assume is called (indirectly) from `htp_tx_res_process_body_data_ex`, does almost the same, except it does reuse the `time_before` value.

Does this mean `duration` will always contain the full time spent?

Does this mean that `time_spent` is actually too high, as both `htp_tx_res_process_body_data_ex` and `htp_tx_res_process_body_data_decompressor_callback` increase it while using the same `time_before` as their base?

Reply:
> Does this mean duration will always contain the full time spent?

No, `duration` contains the duration time for one "atomic" interval between two `gettimeofday`

> Does this mean that time_spent is actually too high, as both htp_tx_res_process_body_data_ex and htp_tx_res_process_body_data_decompressor_callback increase it while using the same time_before as their base?

No because `htp_tx_res_process_body_data_decompressor_callback` updates `time_before`
That is the _trick_